File: autorouter/src/main/java/pt/isel/autorouter/AutoRouter.java
package pt.isel.autorouter;

import pt.isel.autorouter.annotations.AutoRoute;

import java.lang.reflect.Method;
import java.util.*;
import java.util.stream.Stream;


public class AutoRouter {
    Object controller;
    Class<?> controllerClass;
    Stream<Method> controllerMethods;

    public AutoRouter(Object controller) {
        this.controller = controller;
        this.controllerClass = controller.getClass();
        this.controllerMethods = Arrays.stream(controllerClass.getDeclaredMethods()).filter(m ->
                m.isAnnotationPresent(AutoRoute.class) && m.getReturnType() == Optional.class
        );
    }
}

File: autorouter/src/main/kotlin/pt/isel/autorouter/AutoRouter.kt
package pt.isel.autorouter

import pt.isel.autorouter.routers.AutoRouterDynamic
import pt.isel.autorouter.routers.AutoRouterReflect
import pt.isel.autorouter.server.JsonServer
import java.util.stream.Stream

fun Stream<ArHttpRoute>.jsonServer() = JsonServer(this)

fun Any.autorouterReflect() = AutoRouterReflect(this).autorouterReflect()

fun Any.autorouterDynamic() = AutoRouterDynamic(this).autorouterDynamic()